### ISSUE TITLE
Programmatically Disambiguate When to Step `_LRScheduler`

### ISSUE THREAD

**Issue by ntenenz** [pytorch/pytorch#37768]
## 🚀 Feature
A programmatic means (property) to disambiguate when a `_LRScheduler` should be stepped.

## Motivation
Currently, there are a variety of `_LRScheduler`s found in `torch.optim.lr_scheduler`. While the majority should be stepped per epoch (e.g., `LambdaLR`, `MultiplicativeLR`, `StepLR`, etc), some are designed to be stepped per training batch (`CyclicLR`, `OneCycleLR`, etc). Unfortunately, there is no way (programmatically) to identify which category an arbitrary `_LRScheduler` falls into without either maintaining a hard-coded list (😢) or parsing the docstring (:cry:  :cry:). This makes correctly injecting arbitrary `_LRScheduler`s difficult to do in an extensible fashion.

## Pitch
Add two boolean properties to every `_LRScheduler`:
* steps_on_batch
* steps_on_epoch

Currently, neither symbol is used in [`torch.optim.lr_scheduler`](https://pytorch.org/docs/stable/_modules/torch/optim/lr_scheduler.html), so this should be a strictly additive change as far as I'm aware.

## Alternatives
Every scheduler that exists in PyTorch today exclusively steps per batch or per epoch. Therefore, either of the two properties alone would currently satisfy the aforementioned motivation. However, should a scheduler be created that is stepped at both stages, this option would fail. Later adding the second property would be a pseudo-backwards incompatible change as significant amounts of code would likely include `if not sched.step_epoch: ...` to determine if the scheduler should be stepped per batch (or vice versa).

Given the minimal cost of adding both, the flexibility it enables, and the potential to mitigate future challenges, I have opted to propose both properties.

cc @vincentqb

**Comment by ntenenz** [comment]
The PR referenced above is included to clarify the feature request. It lacks docstrings, tests, etc. and is solely intended to facilitate discussion.

**Comment by ntenenz** [comment]
@ngimel @ezyang, given the decision will be largely made based on API and maintenance concerns rather than technical concerns, I was wondering if there's anything I should do to further flesh out the idea to enable an informed discussion.

**Comment by ngimel** [comment]
cc @vincentqb for optimizer API concerns.

**Comment by ntenenz** [comment]
A snippet to illustrate a basic use case. Namely, I'd like to avoid maintaining the duplicate lists which can be error prone.

```python
opt = SGD(model.parameters(), lr=0.1)
dataloader = DataLoader(...)
if scheduler_name == 'cyclic':
    scheduler = CyclicLR(opt, base_lr=0.01, max_lr=0.1)
elif scheduler_name == 'step':
    scheduler = StepLR(opt, 30)
else:
    scheduler = ExponentialLR(opt, 0.01)

for epoch in range(100):
    for batch in dataloader:
        train_batch(...)
        # currently
        if isinstance(scheduler, CyclicLR): scheduler.step()
        # proposed
        # if scheduler.steps_on_batch: scheduler.step()
    validate(...)
    # currently
    if isinstance(scheduler, [StepLR, ExponentialLR]): scheduler.step()
    # proposed
    # if scheduler.steps_on_epoch: scheduler.step()
```

**Comment by vincentqb** [comment]
Thanks for opening the discussion. There are a few ways we could make something like this happen, but I would worry about schedulers that could be "hybrid" and updated at each batch and each epoch as you mentioned.

I'm not agreeing the a new interface yet, but let's see what could be potential ideas.

1. Add two properties `step_on_batch` or `step_on_epoch`.
1. Add one property `step_on_batch`.
1. ~~Parse the docstring.~~
1. Add two new subclasses of `_LRScheduler`: `_LRSchedulerOnBatch` and `_LRSchedulerOnEpoch`.
1. Use `scheduler.step(is_epoch=True)` to tell the scheduler to behave as an epoch or batch scheduler.
1. ~~`torch.optim` as a list of for each containing type of scheduler.~~

One thing I would add is that schedulers like `ReduceLROnPlateau` require a different parameter to `step` and still need to be considered as special cases. How would you factor that in?

**Comment by ntenenz** [comment]
> Thanks for opening the discussion. There are a few ways we could make something like this happen, but I would worry about schedulers that could be "hybrid" and updated at each batch and each epoch as you mentioned.

Completely understand the reluctance to consider them when selecting a design. They don't exist today, so it may be over-engineering. However, with the flag approach, it appeared to be a low-effort way to future-proof ourselves. Regardless, they're not the primary motivation, at least in my mind, for introducing two flags instead of one.

> I'm not agreeing the a new interface yet, but let's see what could be potential ideas.
> 
> 1. Add two properties `step_on_batch` or `step_on_epoch`.

My preferred approach as stated for reasons above. Namely, it's clear, easy to implement, and provides a bit of extra flexibility. Additionally, I find it semantically more clear than the alternatives and the symmetry more appealing than `not scheduler.steps_on_batch`, but that is purely a personal preference. 

> 2. Add one property `step_on_batch`.

Also clear and reasonable, but we implicitly bind ourselves to the "steps on batch end" or "steps on epoch end" scenarios. Stepping on both (or elsewhere) would likely require a backwards-incompatible change. 

> 3. ~Parse the docstring.~

Looks like we agree this is not the best option.

> 4. Add two new subclasses of `_LRScheduler`: `_LRSchedulerOnBatch` and `_LRSchedulerOnEpoch`.

Without introducing class-specific methods, `isinstance(scheduler, _LRSchedulerOnBatch)` is functionally equivalent to the property `steps_on_batch` and ditto for `_LRSchedulerOnEpoch`. Introducing a parent class allows for specialization of the interface whereas leveraging a property, I'd argue, is more clear. If we foresee the need to specialize the interface across _all schedulers of a given type_, the base class approach is likely the right one.

This could also be merged with options 1 or 2 where the property/properties would be implemented in `_LRSchedulerOnBatch` and `_LRSchedulerOnEpoch`. However, if we were to solely implement the properties, the base classes could always be introduced at a later date in a backwards-compatible fashion if this additional level of class hierarchy were deemed unnecessary at this time.

> 5. Use `scheduler.step(is_epoch=True)` to tell the scheduler to behave as an epoch or batch scheduler.

Are you suggesting invoking `scheduler.step` upon completion of each batch and each epoch with the scheduler performing a no-op if appropriate? This would also work, assuming each scheduler set an appropriate default value to ensure backwards compatibility. The downside is that this method is not externally queryable (at least easily).

> 6. ~`torch.optim` as a list of for each containing type of scheduler.~

Looks like we also agree this is not the best option.

> One thing I would add is that schedulers like `ReduceLROnPlateau` require a different parameter to `step` and still need to be considered as special cases. How would you factor that in?

Unless I'm missing something, if using the flag-based approach, the specialization of `.step()` would not impact the design. `ReduceLROnPlateau` would be modified as below (see the sample PR #37770 for a PoC of what I'm suggesting). If option 5 were to be considered however, it may be beneficial to unify the `step` interface and add the `is_epoch` argument in one fell swoop to minimize impact on users.

```python
class ReduceLROnPlateau(object):
    ...
    @property
    def steps_on_batch(self):
        return False

    @property
    def steps_on_epoch(self):
        return True
```

**Comment by vincentqb** [comment]
> > 1. Add one property `step_on_batch`.
> 
> Also clear and reasonable, but we implicitly bind ourselves to the "steps on batch end" or "steps on epoch end" scenarios. Stepping on both (or elsewhere) would likely require a backwards-incompatible change.

How about `step_on` attribute that can be `batch` or `epoch`? and maybe others later?

**Comment by ntenenz** [comment]
Also a very reasonable choice. An `enum.Enum` would be a clean way of succinctly capturing valid options. A list of valid strings would also be possible but would be less amenable to type hinting and static validation. While the `steps_on` attribute would limit the "step on batch and epoch" scenario, this functionality could be recovered (if desired) by using an `enum.Flag` instead of an `enum.Enum` as the attribute type. This change could be made upon attribute introduction or later in what would likely be a backwards compatible change.

**Comment by vincentqb** [comment]
> > 1. Use `scheduler.step(is_epoch=True)` to tell the scheduler to behave as an epoch or batch scheduler.
> 
> Are you suggesting invoking `scheduler.step` upon completion of each batch and each epoch with the scheduler performing a no-op if appropriate? This would also work, assuming each scheduler set an appropriate default value to ensure backwards compatibility. The downside is that this method is not externally queryable (at least easily).

A variation: we could have two new functions
```
scheduler.step_on_epoch(...)
scheduler.step_on_batch(...)
```
and each scheduler would simply call the step function in either one, while the other would be potentially a no-op.

This is backward compatible, but increase the API.

**Comment by ntenenz** [comment]
The approaches being discussed largely fall into two, not necessarily exclusive, camps.
1. Caller is aware when step should be invoked and acts accordingly
2. Caller is unaware when step should be invoked with no-ops added to unify the API

The latter removes logic from the caller, which may be simpler. However, in this scenario, the caller also lacks any insight into when "real" steps are occuring. Such visibility may be useful in a distributed setting to avoid expensive inter-node syncs as well as other potential scenarios. As alluded to earlier, the `steps_on` property and `step_on_{batch,epoch}` methods would provide the best of both worlds, but at the cost of a more significant API expansion.

At this point, my personal preference would be to add the `steps_on` property (returning an enum) and deferring on the multiple step methods, which could always be added later if there were sufficient demand. That being said, I'm far from a core contributor so I defer to your judgement. 🙂

**Comment by vincentqb** [comment]
I do agree with explicit is better, and leaving the control in the hand of the users. Alright, let's go for a `steps_on` property that can be either "batch" or "epoch". If you like this, do you want to update your pull request for this?

**Comment by ntenenz** [comment]
@vincentqb, I've updated the PR to highlight the general idea.

**Comment by ntenenz** [comment]
bump @vincentqb

**Comment by vincentqb** [comment]
Responded: one element that came up while looking at the PR is what to do with `LambdaLR` and `MultiplicativeLR` which depend on what the user would like to do :)

**Comment by ntenenz** [comment]
Currently the documentation for [`LambdaLR`](https://pytorch.org/docs/stable/optim.html#torch.optim.lr_scheduler.LambdaLR) and ['MultiplicativeLR'](https://pytorch.org/docs/stable/optim.html#torch.optim.lr_scheduler.MultiplicativeLR) suggests they should be stepped every epoch -- at least that's my (perhaps faulty) interpretation based on the attached examples and param `last_epoch`. While one could certainly step them arbitrarily, I'll leave it up to you whether that should be a supported use case or one which is nonstandard but also not precluded.

**Comment by harryhan618** [comment]
@ntenenz  Hi! I'm thinking a `Compose` wrapper for lr_scheduler like `torchvision.transforms.Compose` in issue #41307, so that different scheduler strategy can be combined and used at different time. I think this is somehow related to my idea. Any suggestions? I think I'm gonna PR a first version these days.

**Comment by ntenenz** [comment]
I suspect this PR would solve one particularly thorny issue you'd face. Notably, right now there's no easy/clean way to determine when a LRScheduler should be stepped. This PR would add a unified interface allowing any scheduler to signify (potentially dynamically) when to step.

**Comment by harryhan618** [comment]
@ntenenz Hi. I'm thinking a Compose wrapper so that we can call `step` in a neat way, like this:
``` python
"""
we may have a bunch of lr_schedulers, with warmup strategy, one lr_scheduler in the middle of training, another in the end, another by epoch. Or, some lr_schedulers for one part of the model during some time, others on other part of the model, etc.
"""
lr_comp = LRCompose_Init_API   # a list of lr_schedulers, either by epoch or by iteration, also with their start time and end time
...
for x, y in dataloader:
    ...
    optim.step()
    lr_comp.step()
lr_comp.step()
```
We only need to call `step` twice, once by iteration, once by epoch. **No need to judge if lr_scheduler is by iteration or by epoch explicitly for users.**
How about this?

**Comment by ntenenz** [comment]
@vincentqb, thoughts on the current status of this PR?

**Comment by harryhan618** [comment]
@ntenenz Thanks for you reply. But in your PR, your have hard assign some lr_schedulers to be step by epoch and others by iteration. That doesn't make sense. For example, sometimes I would use `ExponentialLR` by epoch, somtimes by iteration. So by your design, another `setter` property should also be added to let user have their own choice. If a user want to implement a new lr_scheduler, not only does user have to implement `get_lr`, the user also has to remember to add these two method. So I don't think it's a elegant and flexible design.

So why don't we have a Compse like `torchvision.transforms.Compose`? We can let user to determine which lr_scheduler is stepped by iteration or by epoch.

**Comment by ntenenz** [comment]
> @ntenenz Thanks for you reply. But in your PR, your have hard assign some lr_schedulers to be step by epoch and others by iteration. That doesn't make sense. For example, sometimes I would use `ExponentialLR` by epoch, somtimes by iteration. So by your design, another `setter` property should also be added to let user have their own choice. If a user want to implement a new lr_scheduler, not only does user have to implement `get_lr`, the user also has to remember to add these two method. So I don't think it's a elegant and flexible design.
> 
> So why don't we have a Compse like `torchvision.transforms.Compose`? We can let user to determine which lr_scheduler is stepped by iteration or by epoch.

@harryhan618, While many LR schedules can be used both per batch and per epoch, as defined today, the code in the corresponding PR aligns with the existing documentation. For instance, the docs define [`MultiplicativeLR`](https://pytorch.org/docs/1.5.1/optim.html#torch.optim.lr_scheduler.MultiplicativeLR), [`StepLR`](https://pytorch.org/docs/1.5.1/optim.html#torch.optim.lr_scheduler.StepLR), and [`ExponentialLR`](https://pytorch.org/docs/1.5.1/optim.html#torch.optim.lr_scheduler.ExponentialLR) to step per epoch both in the parameter documentation (`last_epoch`) and the code examples. Conversely, the three LR schedulers defined to step per batch in the PR ([`CyclicLR`](https://pytorch.org/docs/1.5.1/optim.html#torch.optim.lr_scheduler.CyclicLR), [`OneCycleLR`](https://pytorch.org/docs/1.5.1/optim.html#torch.optim.lr_scheduler.OneCycleLR), and [`CosineAnnealingWarmRestarts`](https://pytorch.org/docs/1.5.1/optim.html#torch.optim.lr_scheduler.CosineAnnealingWarmRestarts)) explicitly define the parameter `last_epoch` to be the index of the last batch (not epoch), provide a code sample stepping per batch, or both. 

I agree that providing a user the flexibility to modify LR schedule behavior would be nice but would argue it may be out of scope for this PR given it would be further additional functionality that may require some thought (e.g., what happens if a user tries to change the behavior of a scheduler that cannot function in the new mode?). Nothing in this change is backwards-incompatible (meaning existing use cases of a per epoch schedule being used per batch are still supported), and future changes can be made in a backwards-compatible fashion. Furthermore, a user, if desired, can override the behavior for your use case by inheriting from the schedule and simply define `steps_on` to return the desired value.

**Comment by vincentqb** [comment]
Thanks for bumping the issue :)

After thinking more about the issue and an offline chat with @fmassa, I realize this issue is quite related to the fact that pytorch doesn't have a training loop builtin with optimizer/scheduler stepping. We want the code to be quite flexible, and leave the control in the hand of the user. The distinction between iteration and epoch and whether we should step over one or the other depends on the particular application and there is often many ways of organizing things:, e.g. (1) what if we decide to not even have an epoch loop, but just a data loader loop that also repeats the data?) (2) are  `LambdaLR` and `MultiplicativeLR` stepped by epoch or iterations? (mentioned in [comment](https://github.com/pytorch/pytorch/pull/37770/files#r445887742))

Most schedulers can also just be scaled to epoch or iterations irrespective of the vocabulary used in the documentation. An alternative solution to what is suggested [above](https://github.com/pytorch/pytorch/issues/37768#issuecomment-627390637) is simply improving documentation with how to illustrate scaling them?

**Comment by ntenenz** [comment]
> Thanks for bumping the issue :)
> 
> After thinking more about the issue and an offline chat with @fmassa, I realize this issue is quite related to the fact that pytorch doesn't have a training loop builtin with optimizer/scheduler stepping. We want the code to be quite flexible, and leave the control in the hand of the user. The distinction between iteration and epoch and whether we should step over one or the other depends on the particular application and there is often many ways of organizing things:, e.g. (1) what if we decide to not even have an epoch loop, but just a data loader loop that also repeats the data?) (2) are `LambdaLR` and `MultiplicativeLR` stepped by epoch or iterations? (mentioned in [comment](https://github.com/pytorch/pytorch/pull/37770/files#r445887742))
> 
> Most schedulers can also just be scaled to epoch or iterations irrespective of the vocabulary used in the documentation. An alternative solution to what is suggested [above](https://github.com/pytorch/pytorch/issues/37768#issuecomment-627390637) is simply improving documentation with how to illustrate scaling them?

@vincentqb, I agree with all of the above points. However, I'd argue the motivation for this change isn't necessarily to standardize how a scheduler is used though but rather to indicate to the caller (i.e., training loop) how it should be used in that particular use case.

For instance, one could imagine making the LR schedule type a hyperparameter that is searched over or configured via an external config file. Unfortunately, in either case, there would be no way for the training loop to know when to step (other than to explicitly condition on scheduler type), creating a tight coupling between the scheduler and training loop. By using `steps_on`, this coupling is managed by the scheduler property, for which this PR would provide sensible defaults and users could readily override.

To answer your two questions directly, in both (1) and (2), I'd expect the user to override the `steps_on` property dynamically, tracking batch count internally if needed.

If you feel this functionality doesn't warrant inclusion in the PyTorch at this time, I'd be happy to close this issue/PR and maintain the code externally, however, synchronizing the two would incur higher maintenance costs, so I'd obviously prefer for it be included here. Just let me know either way. 🙂

**Comment by vincentqb** [comment]
Thanks for the input :) Indeed, after thinking more about this, I realize this will add to the complexity, and the new feature wouldn't satisfy everyone. Most schedulers can actually be converted as "step on each iterations" with corresponding scaling of the parameters, so a simple code can be written where `step` is always done on each iteration, if the user so desires. Thanks again for the suggestion! I'll close the issue, but feel free to re-open if you would like to continue the discussion :)
